pyqtgraph/graphicsItems/TargetItem.py

```diff
@@ -1,132 +1,352 @@
-    
-    def dataBounds(self, axis, frac=1.0, orthoRange=None):
-        return [0, 0]
+        return self.shape().boundingRect()
 
-    def viewTransformChanged(self):
-        self._picture = None
-        self.prepareGeometryChange()
-        self._updateLabel()
+    def paint(self, p, *_):
+        p.setPen(self.currentPen)
+        p.setBrush(self.currentBrush)
+        p.drawPath(self.shape())
 
-    def _updateLabel(self):
-        if self.label is None:
-            return
+    def setPath(self, path):
+        if path != self._path:
+            self._path = path
+            self._shape = None
+        return None
 
-        # find an optimal location for text at the given angle
-        angle = self.labelAngle * np.pi / 180.
-        lbr = self.label.boundingRect()
-        center = lbr.center()
-        a = abs(np.sin(angle) * lbr.height()*0.5)
-        b = abs(np.cos(angle) * lbr.width()*0.5)
-        r = max(self._radii) + 2 + max(a, b)
-        pos = self.mapFromScene(self.mapToScene(QtCore.QPointF(0, 0)) + r * QtCore.QPointF(np.cos(angle), -np.sin(angle)) - center)
-        self.label.setPos(pos)
-
-    def paint(self, p, *args):
-        if self._picture is None:
-            self._drawPicture()
-        self._picture.play(p)
-
-    def _drawPicture(self):
-        self._picture = QtGui.QPicture()
-        p = QtGui.QPainter(self._picture)
-        p.setRenderHint(p.Antialiasing)
-        
-        # Note: could do this with self.pixelLength, but this is faster.
-        o = self.mapToScene(QtCore.QPointF(0, 0))
-        dx = (self.mapToScene(QtCore.QPointF(1, 0)) - o).x()
-        dy = (self.mapToScene(QtCore.QPointF(0, 1)) - o).y()
-        if dx == 0 or dy == 0:
-            p.end()
-            self._bounds = QtCore.QRectF()
-            return
-        px = abs(1.0 / dx)
-        py = abs(1.0 / dy)
-        
-        r, w, h = self._radii
-        w = w * px
-        h = h * py
-        rx = r * px
-        ry = r * py
-        rect = QtCore.QRectF(-rx, -ry, rx*2, ry*2)
-        p.setPen(self.pen)
-        p.setBrush(self.brush)
-        p.drawEllipse(rect)
-        p.drawLine(Point(-w, 0), Point(w, 0))
-        p.drawLine(Point(0, -h), Point(0, h))
-        p.end()
-        
-        bx = max(w, rx)
-        by = max(h, ry)
-        self._bounds = QtCore.QRectF(-bx, -by, bx*2, by*2)
+    def shape(self):
+        if self._shape is None:
+            s = self.generateShape()
+            if s is None:
+                return self._path
+            self._shape = s
+            # beware--this can cause the view to adjust
+            # which would immediately invalidate the shape.
+            self.prepareGeometryChange()
+        return self._shape
+
+    def generateShape(self):
+        dt = self.deviceTransform()
+        if dt is None:
+            self._shape = self._path
+            return None
+        v = dt.map(QtCore.QPointF(1, 0)) - dt.map(QtCore.QPointF(0, 0))
+        va = atan2(v.y(), v.x())
```

Comment:
I do not understand what this set of calculations does. The fact that is uses both ArcTan and inverse transforms seems a little suspicious to me. This is the only occurrrence of atan across the entire codebase. :)

Reply:
honestly this bit was inherited from the original PR; i probably should sketch out the math a bit more closely

Reply:
I re-arranged the device transformation math in hopes to give it a bit more context.

```python
        tr = QtGui.QTransform()
        tr.scale(self.scale, self.scale)
        
        devPos = dt.map(QtCore.QPointF(0, 0))
        tr.translate(devPos.x(), devPos.y())
        
        v = dt.map(QtCore.QPointF(1, 0)) - dt.map(QtCore.QPointF(0, 0))
        va = atan2(v.y(), v.x())
        tr.rotate(va * 180.0 / pi)
        
        dti = fn.invertQTransform(dt)
        return dti.map(tr.map(self._path))
```

I'm interpreting this to handle the use-case of the coordinate system being rotated with respect to the screen.  I'm not sure pyqtgraph even supports that functionality!

Reply:
also we do use `np.arctan2` in other parts of the library for this same bit in `ROI.py`.